librad/src/id.rs

```diff
@@ -0,0 +1,273 @@
+            ));
+        }
+        Ok(())
+    }
+
+    fn check_key(
+        &self,
+        key: &PublicKey,
+        by: &Signatory,
+        resolver: &impl Resolver<User>,
+    ) -> Result<(), Error> {
+        match by {
+            Signatory::OwnedKey => {
+                if !self.has_key(key) {
+                    return Err(Error::KeyNotPresent(key.to_owned()));
+                }
+            },
+            Signatory::User(uri) => {
+                let user = resolver.resolve(&uri)?;
+                if !user.has_key(key) {
+                    return Err(Error::UserKeyNotPresent(uri.to_owned(), key.to_owned()));
+                }
+            },
+        }
+        Ok(())
+    }
+
+    fn compute_signature(&self, key: &Key) -> Result<Signature, Error> {
+        Ok(key.sign(&self.canonical_data()?))
+    }
+
+    fn sign(
+        &mut self,
+        key: &Key,
+        by: &Signatory,
+        resolver: &impl Resolver<User>,
+    ) -> Result<(), Error> {
+        let public_key = key.public();
+        self.check_key(&public_key, by, resolver)?;
+        let signature = EntitySignature {
+            by: by.to_owned(),
+            sig: self.compute_signature(key)?,
+        };
+        self.signatures_mut().insert(public_key, signature);
+        Ok(())
+    }
+
+    fn check_signature(
+        &mut self,
+        key: &PublicKey,
+        by: &Signatory,
+        signature: &Signature,
+        resolver: &impl Resolver<User>,
+    ) -> Result<(), Error> {
+        self.check_key(key, by, resolver)?;
+        if signature.verify(&self.canonical_data()?, key) {
+            Ok(())
+        } else {
+            Err(Error::SignatureVerificationFailed)
+        }
+    }
+}
+
+pub struct User {
+    pub name: String,
+    pub revision: u64,
+
+    pub devices: HashSet<PeerId>,
+
+    pub uri: RadicleUri,
+    pub signatures: HashMap<PublicKey, EntitySignature>,
+}
+
+impl Entity for User {
+    fn name(&self) -> &str {
+        &self.name
+    }
+    fn revision(&self) -> u64 {
+        self.revision
+    }
+
+    fn canonical_data(&self) -> Result<Vec<u8>, Error> {
+        unimplemented!()
+    }
+
+    fn uri(&self) -> &RadicleUri {
+        &self.uri
+    }
+    fn signatures(&self) -> &HashMap<PublicKey, EntitySignature> {
+        &self.signatures
+    }
+    fn signatures_mut(&mut self) -> &mut HashMap<PublicKey, EntitySignature> {
+        &mut self.signatures
+    }
+
+    fn keys_count(&self) -> usize {
+        self.devices.len()
+    }
+    fn has_key(&self, key: &PublicKey) -> bool {
+        self.devices.contains(&PeerId::from(key.to_owned()))
+    }
+    fn keys<'a>(&'a self) -> Box<dyn Iterator<Item = PublicKey> + 'a> {
+        Box::new(self.devices.iter().map(|id| id.device_key().to_owned()))
+    }
+
+    fn certifiers_count(&self) -> usize {
+        0
+    }
+    fn has_certifier(&self, _key: &RadicleUri) -> bool {
+        false
+    }
+    fn certifiers(&self) -> Box<dyn Iterator<Item = RadicleUri>> {
```

Comment:
What are `certifiers`? Does it make sense to live in `Entity` and have these implementations for `User`? To me it looks like a code smell :thinking: